contracts/optimistic-ethereum/OVM/OVM_BondManager.sol

```diff
@@ -0,0 +1,223 @@
+            bond.withdrawalTimestamp <= timestamp + disputePeriodSeconds &&
+            bond.withdrawing >= requiredCollateral
+        ) {
+            bond.withdrawing -= requiredCollateral;
+        }
+    }
+
+    /// Sequencers call this function to post collateral which will be used for
+    /// the `appendBatch` call
+    function deposit(uint256 amount) public {
+        require(
+            token.transferFrom(msg.sender, address(this), amount),
+            Errors.ERC20_ERR
+        );
+
+        // This cannot overflow
+        bonds[msg.sender].locked += amount;
+    }
+
+    /// Starts the withdrawal for a publisher
+    function startWithdrawal() public {
+        Bond storage bond = bonds[msg.sender];
+        require(bond.withdrawalTimestamp == 0, Errors.WITHDRAWAL_PENDING);
+        require(bond.locked >= requiredCollateral, Errors.NOT_ENOUGH_COLLATERAL);
+        bond.locked -= requiredCollateral;
+
+        bond.withdrawing += requiredCollateral;
+        bond.withdrawalTimestamp = block.timestamp;
+    }
+
+    /// Finalizes a pending withdrawal from a publisher
+    function finalizeWithdrawal() public {
+        Bond storage bond = bonds[msg.sender];
+
+        require(
+            block.timestamp >= bond.withdrawalTimestamp + disputePeriodSeconds, 
+            Errors.TOO_EARLY
+        );
+        require(bond.withdrawing >= requiredCollateral, Errors.SLASHED);
+        bond.withdrawing -= requiredCollateral;
+        bond.withdrawalTimestamp = 0;
+        
+        require(
+            token.transfer(msg.sender, requiredCollateral),
+            Errors.ERC20_ERR
+        );
+    }
+
+    /// Claims the user's reward for the witnesses they provided
+    function claim(bytes32 _preStateRoot) public {
+        Rewards storage rewards = witnessProviders[_preStateRoot];
+
+        // only allow claiming if fraud was proven in `finalize`
+        require(rewards.canClaim, Errors.CANNOT_CLAIM);
+
+        // proportional allocation - only reward 50% (rest gets locked in the
+        // contract forever
+        uint256 amount = (requiredCollateral * rewards.numClaims[msg.sender]) / (2 * rewards.total);
+
+        // reset the user's claims so they cannot double claim
+        rewards.numClaims[msg.sender] = 0;
+
+        // transfer
+        require(token.transfer(msg.sender, amount), Errors.ERC20_ERR);
+    }
+
+    /// Sets the required collateral for posting a state root
+    /// Callable only by the contract's deployer.
+    function setRequiredCollateral(uint256 newValue) public {
+        require(newValue > requiredCollateral, Errors.LOW_VALUE);
+        require(msg.sender == owner, Errors.ONLY_OWNER);
+        requiredCollateral = newValue;
+    }
+
+    /// Checks if the user is collateralized for the batchIndex
+    function isCollateralized(address who) public view returns (bool) {
+        require(bonds[who].locked >= requiredCollateral, Errors.NOT_ENOUGH_COLLATERAL);
+        return true;
```

Comment:
eh I prefer it this way, similarly to how most ERC20's transfers return true but revert internally, instead of returning false on error conditions. It just removes the burden from the consumer calling this function